File: src/main/java/org/stormgears/powerup/subsystems/elevatorclimber/ElevatorSharedTalons.java
package org.stormgears.powerup.subsystems.elevatorclimber;

import com.ctre.phoenix.motorcontrol.ControlMode;
import org.stormgears.powerup.TalonIds;
import org.stormgears.utils.decoupling.ITalon;
import org.stormgears.utils.decoupling.TalonFactoryKt;

public class ElevatorSharedTalons {
	private static ElevatorSharedTalons instance;

	public static ElevatorSharedTalons getInstance() {
		return instance;
	}

	public static final int TALON_FPID_TIMEOUT = 10;

	private static final int MASTER_MOTOR_TALON_ID = TalonIds.ELEVATOR_MASTER;
	private static final int SLAVE_MOTOR_TALON_ID = TalonIds.ELEVATOR_SLAVE;

	private ITalon masterMotor;
	private ITalon slaveMotor;

	private ElevatorSharedTalons(int masterMotorId, int slaveMotorId) {
		System.out.println("Initializing elevator talons");

		masterMotor = TalonFactoryKt.createTalon(masterMotorId);
		slaveMotor = TalonFactoryKt.createTalon(slaveMotorId);
		slaveMotor.set(ControlMode.Follower, masterMotorId);

		invert(true);
		masterMotor.setSensorPhase(true);

		masterMotor.configPeakCurrentLimit(30, TALON_FPID_TIMEOUT);
		masterMotor.configContinuousCurrentLimit(60, TALON_FPID_TIMEOUT);
		masterMotor.enableCurrentLimit(true);
		slaveMotor.configPeakCurrentLimit(30, TALON_FPID_TIMEOUT);
		slaveMotor.configContinuousCurrentLimit(60, TALON_FPID_TIMEOUT);
		slaveMotor.enableCurrentLimit(true);
	}

	void invert(boolean inverted) {
		masterMotor.setInverted(inverted);
		slaveMotor.setInverted(inverted);
	}

	public static void init() {
		instance = new ElevatorSharedTalons(MASTER_MOTOR_TALON_ID, SLAVE_MOTOR_TALON_ID);
	}

	public ITalon getMasterMotor() {
		return masterMotor;
	}

	public ITalon getSlaveMotor() {
		return slaveMotor;
	}
}

File: src/main/java/org/stormgears/powerup/subsystems/elevatorclimber/ElevatorSharedTalons.kt
package org.stormgears.powerup.subsystems.elevatorclimber

import com.ctre.phoenix.motorcontrol.*
import org.stormgears.powerup.Robot
import org.stormgears.utils.talons.FactoryTalonConfig
import org.stormgears.utils.talons.ITalon
import org.stormgears.utils.talons.LocalLimitSwitchSourceConfig
import org.stormgears.utils.talons.createTalon

object ElevatorSharedTalons {
	private val MASTER_MOTOR_TALON_ID = Robot.config.elevatorMasterTalonId
	private val SLAVE_MOTOR_TALON_ID = Robot.config.elevatorSlaveTalonId

	val masterMotor: ITalon
	val slaveMotor: ITalon

	open class ElevatorTalonConfig : FactoryTalonConfig() {
		// Raise
		override val profileSlot0 = object : DefaultPIDSlot() {
			override val kP = Robot.config.elevatorRaiseP
			override val kI = Robot.config.elevatorRaiseI
			override val kD = Robot.config.elevatorRaiseD
			override val kF = 0.0
		}

		// Lower
		override val profileSlot1 = object : DefaultPIDSlot() {
			override val kP = Robot.config.elevatorLowerP
			override val kI = Robot.config.elevatorLowerI
			override val kD = Robot.config.elevatorLowerD
			override val kF = 0.0
		}

		override val neutralMode = NeutralMode.Brake
		override val inverted = true
		override val sensorPhase = true

		override val peakCurrentLimit = 30
		override val continuousCurrentLimit = 60
		override val enableCurrentLimit = true

		override val forwardLimitSwitchSource = LocalLimitSwitchSourceConfig(LimitSwitchSource.FeedbackConnector, LimitSwitchNormal.NormallyOpen)
		override val reverseLimitSwitchSource = LocalLimitSwitchSourceConfig(LimitSwitchSource.FeedbackConnector, LimitSwitchNormal.NormallyOpen)

		override val statusFramePeriod = mapOf(
			StatusFrameEnhanced.Status_13_Base_PIDF0 to 10,
			StatusFrameEnhanced.Status_3_Quadrature to 10
		)
	}

	class MasterElevatorTalonConfig : ElevatorTalonConfig() {
//		override val forwardSoftLimitThreshold = -1145000
//		override val forwardSoftLimitEnable = true
	}

	val elevatorTalonConfig = ElevatorTalonConfig()
	val masterElevatorTalonConfig = MasterElevatorTalonConfig()

	init {
//		println("Initializing elevator talons")

		masterMotor = createTalon(MASTER_MOTOR_TALON_ID)
		slaveMotor = createTalon(SLAVE_MOTOR_TALON_ID)

//		TalonDebugger(arrayOf(masterMotor, slaveMotor)).start()


		masterMotor.setConfig(masterElevatorTalonConfig)
		slaveMotor.setConfig(elevatorTalonConfig)

		slaveMotor.set(ControlMode.Follower, MASTER_MOTOR_TALON_ID.toDouble())
	}
}
